src/qibo/validation/gate_set_tomography.py

```diff
@@ -0,0 +1,2089 @@
+            circ.add(gates.H(0))
+
+        elif k == 9:  # |+> |1>
+            circ = Circuit(nqubits, density_matrix=True)
+            circ.add(gates.H(0))
+            circ.add(gates.X(1))
+
+        elif k == 10:  # |+> |+>
+            circ = Circuit(nqubits, density_matrix=True)
+            circ.add(gates.H(0))
+            circ.add(gates.H(1))
+
+        elif k == 11:  # |+> |y+>
+            circ = Circuit(nqubits, density_matrix=True)
+            circ.add(gates.H(0))
+            circ.add(gates.H(1))
+            circ.add(gates.S(1))
+
+        # --------------------------------------------
+
+        elif k == 12:  # |y+> |0>
+            circ = Circuit(nqubits, density_matrix=True)
+            circ.add(gates.H(0))
+            circ.add(gates.S(0))
+
+        elif k == 13:  # |y+> |1>
+            circ = Circuit(nqubits, density_matrix=True)
+            circ.add(gates.H(0))
+            circ.add(gates.S(0))
+            circ.add(gates.X(1))
+
+        elif k == 14:  # |y+> |+>
+            circ = Circuit(nqubits, density_matrix=True)
+            circ.add(gates.H(0))
+            circ.add(gates.S(0))
+            circ.add(gates.H(1))
+
+        elif k == 15:  # |y+> |y+>
+            circ = Circuit(nqubits, density_matrix=True)
+            circ.add(gates.H(0))
+            circ.add(gates.S(0))
+            circ.add(gates.H(1))
+            circ.add(gates.S(1))
+
+        # --------------------------------------------
+
+        # If no class_qubit_gate
+        if class_qubit_gate is None:
+            pass
+
+        # If class_qubit_gate is present (with/without theta)
+        elif class_qubit_gate is not None:
+            if theta is None:
+                print(f"Gate set tomography for 2-qubit gate: {class_qubit_gate}.")
+                circ.add(getattr(gates, class_qubit_gate)(targ_qb[0], ctrl_qb[0]))
+            elif theta is not None:
+                print(
+                    f"Gate set tomography for 2-qubit gate: {class_qubit_gate}({theta})."
+                )
+                circ.add(
+                    getattr(gates, class_qubit_gate)(targ_qb[0], ctrl_qb[0], theta)
+                )
+
+        # --------------------------------------------
+
+        # initial_state = np.zeros((2**nqubits))
+        # initial_state[0] = 1
+        # ================= START OF 16 MEASUREMENT BASES =================|||
+        for j in range(0, 16):
+            # print('GST 2qb blank circ: initial state k=%d, measurement basis %d' %(k,j))
+
+            if j == 0:  # Identity basis  # Identity basis
+                circ2 = Circuit(nqubits, density_matrix=True)
+                new_circ = circ + circ2
+                if noise_model is not None and backend.name != "qibolab":
+                    new_circ = noise_model.apply(new_circ)
+                new_circ.add(gates.M(0, basis=gates.Z))
+                new_circ.add(gates.M(1, basis=gates.Z))
+
+                expectation_val = GST_measure_2qb(new_circ, k, j)
```

Comment:
These blocks are all identical except for the basis you use to measure, if I am not mistaken, better to have a function that takes as input the measurement basis and computes the expectation value

Reply:
Thanks Andrea, I will look into building a helper function for j. Thanks for the suggestion, it will definitely streamline the code.

Reply:
This is done too.